src/cargo/ops/cargo_new.rs

```diff
@@ -412,74 +510,84 @@ fn mk(config: &Config, opts: &MkOptions) -> CargoResult<()> {
+    try!(walk_template_dir(&template_path, &mut |entry| {
+        let entry_path = entry.path();
+        let entry_str = entry_path.to_str().unwrap();
+        let template_dir_str = template_path.to_str().unwrap();
+
+        // the path we have here is the absolute path to the file in the template directory
+        // we need to trim this down to be just the path from the root of the template.
+        // For example:
+        //    /home/user/.cargo/templates/foo/Cargo.toml => Cargo.toml
+        //    /home/user/.cargo/templates/foo/src/main.rs => src/main.rs
+        //    C:\user\.cargo\templates\foo\src\main.rs => src/main.rs
+        let mut dest_file_name = entry_str.replace(template_dir_str, "");
+        if dest_file_name.starts_with(MAIN_SEPARATOR) {
+            dest_file_name.remove(0);
+        }
 
-        if let Some(src_dir) = path_of_source_file.parent() {
-            try!(fs::create_dir_all(src_dir));
+        let mut template_str = String::new();
+        File::open(&entry_path).unwrap().read_to_string(&mut template_str).unwrap();
+        let mut dest_path = PathBuf::from(path).join(dest_file_name);
+
+        // dest_file_name could now refer to a file inside a directory which doesn't yet exist
+        // to figure out if this is the case, get all the components in the dest_file_name and check
+        // how many there are. Files in the root of the new project direcotory will have two
+        // components, anything more than that means the file is in a sub-directory, so we need
+        // to create it.
+        {
+            let components  = dest_path.components().collect::<Vec<_>>();
+            if components.len() > 2 {
+                if let Some(p) = dest_path.parent() {
+                    if fs::metadata(&p).is_err() {
+                        let _ = fs::create_dir_all(&p);
+                    }
+                }
+            }
         }
 
-        let default_file_content : &[u8] = if i.bin {
-            b"\
-fn main() {
-    println!(\"Hello, world!\");
-}
-"
-        } else {
-            b"\
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-    }
-}
-"
-        };
+        // if we are running init, we do not clobber existing files.
+        if fs::metadata(&dest_path).is_ok() {
+            return;
+        }
 
-        if !fs::metadata(&path_of_source_file).map(|x| x.is_file()).unwrap_or(false) {
-            try!(paths::write(&path_of_source_file, default_file_content));
+        // if the template file has the ".hbs" extension, remove that to get the correct
+        // name for the generated file
+        if let Some(ext) = dest_path.clone().extension() {
+            if ext.to_str().unwrap() == "hbs" {
+                dest_path.set_extension("");
+            }
         }
-    }
+
+        // create the new file & render the template to it
+        let mut dest_file = OpenOptions::new().write(true).create(true).open(&dest_path).unwrap();
```

Comment:
This `unwrap` should be an error (I think there may also be a utility function for this)